Schedulers contract update after security review (#669)
rsksmart/devportal: rif/scheduler/contracts.md

The plans must specify the:
- Price per execution,

> - The time window (in seconds) within a scheduled execution should run.

Reviewer: ```suggestion
- The time window (in seconds) within a scheduled execution should run.
- The maximum amount of gas the internal call can spend
**The price of the plan strongly depends on these last two values**
```

> We recommend a value more than 15 seconds.

Reviewer: ```suggestion
We recommend a value more than 15 seconds. This value will depend strongly on the [service](./services) configuration.
```

> We recommend a value bigger than 15 seconds.

Reviewer 2: ```suggestion
We recommend a value more than 15 seconds.
```

The plans must specify the:
- Price per execution,
- The time window (in seconds) within a scheduled execution should run.
- The maximum amount of gas the internal call can spend
**The price of the plan strongly depends on these last two values**

> We recommend a value more than 15 seconds. This value will depend strongly on the [service](./services) configuration.

Reviewer 2: Looks like we're missing the file `rif/scheduler/services.md` - can you either commit this file, or remove the link to it?

CI build failure:

```
- ./_site/rif/scheduler/contracts/index.html
  *  internally linking to ./services, which does not exist (line 1980)
     <a href="./services">service</a>
rake aborted!
HTML-Proofer found 1 failure!
```

Reviewer: It is there 

https://github.com/rsksmart/devportal/blob/schedulers-contract-update-security-review/rif/scheduler/services.md

Reviewer 2: :thinking: just re-ran CI, seems to be failing for the same reason

This performs execution in this way:

```solidity
(bool success, bytes memory result) = payable(execution.to).call{ gas: plan.gas, value: execution.value }(execution.data);
```

After execution is completed, the Service Provider will receive
the payment for the execution.

If the transaction is submitted before time window, the execution will be rejected and it will remain scheduled.
If it is submitted after the time window, it won't be executed and its state will be Overdue.

In any case the Service Provider is not responsible for the successful execution of the submitted transaction. It will only guarantee that it is submitted as scheduled.

> ### Overdue executions refunding

Reviewer 2: ```suggestion
### Refunding overdue executions

```

The requestor can request the refund of the overdue scheduled transactions (those that were not executed on time, inside the execution window).

> To clain the refund, the requestor should call the following method:

Reviewer 2: ```suggestion
To claim the refund, the requestor should call the following method:
```

```solidity
  function requestExecutionRefund(bytes32 id) external
```

This will set the requested execution as refunded and the remaining executions for thar plan will be increased by one. Also if there was some value transferred during the scheduling, it will be returned to the requestor.

> ### Plans refunding

Reviewer 2: ```suggestion
### Refunding plans

```

3. Then, each requestor should call the method `requestPlanRefund(uint256 plan)` for each plan that they purchased to receive a transfer for the value of the remaining balance. Please notice that this is only available while the contract is paused.

Steps 2 and 3 can be repeated as many times as needed. These can also be combined in a single contract call using the `multicall` method.

> ### Multicall
> The contract allows to combine multiple call in a single one using the methods.

Reviewer 2: ```suggestion
### Multicall

The contract allows one to combine multiple calls into a single one using these methods.
```

3. Then, each requestor should call the method `requestPlanRefund(uint256 plan)` for each plan that they purchased to receive a transfer for the value of the remaining balance. Please notice that this is only available while the contract is paused.

Steps 2 and 3 can be repeated as many times as needed. These can also be combined in a single contract call using the `multicall` method.

### Multicall

The contract allows one to combine multiple calls into a single one using this method.

```solidity 
multicall(bytes[] calldata data, bool revertIfFails)
```

This method receives:
- calldata: an array of encoded contract calls to be executed

> - revertIfFails: if true when a transaction in calldata is executed and it reverts, all the transactions will be reverted. Otherwise, it will continue executing the following transactions.

Reviewer: ```suggestion
- revertIfFails: if true when a transaction in calldata is executed and it reverts, all the transactions will be reverted. Otherwise, it will continue executing the following transactions.

> WARNING: using `revertIfFails` set to `true` can incur in DoS attackas: by spamming one transaction, all the others will fail. In the other hand, using `revertIfFails` set to `false` will only allow to retrieve the index of the calls that failed.
```
